permissions/resource_permissions.go

```diff
@@ -108,315 +28,137 @@ type PermissionsAPI struct {
+	idParts := strings.Split(objectID, "/")
+	id := idParts[len(idParts)-1]
+	withOwner := mapping.addOwnerPermissionIfNeeded(objectACL, ownerOpt)
+	_, err = w.Permissions.Set(a.context, iam.PermissionsRequest{
+		RequestObjectId:   id,
+		RequestObjectType: mapping.requestObjectType,
+		AccessControlList: withOwner,
+	})
+	if err != nil {
+		if strings.Contains(err.Error(), "with no existing owner must provide a new owner") {
+			_, err = w.Permissions.Set(a.context, iam.PermissionsRequest{
+				RequestObjectId:   id,
+				RequestObjectType: mapping.requestObjectType,
+				AccessControlList: objectACL,
+			})
 		}
-		return warehouse.CreatorName, nil
+		return err
 	}
-	return "", nil
+	return nil
 }
 
-func (a PermissionsAPI) ensureCurrentUserCanManageObject(objectID string, objectACL AccessControlChangeList) (AccessControlChangeList, error) {
-	if !a.shouldExplicitlyGrantCallingUserManagePermissions(objectID) {
-		return objectACL, nil
-	}
+func (a PermissionsAPI) getCurrentUser() (string, error) {
 	w, err := a.client.WorkspaceClient()
 	if err != nil {
-		return objectACL, err
+		return "", err
 	}
 	me, err := w.CurrentUser.Me(a.context)
 	if err != nil {
-		return objectACL, err
+		return "", err
 	}
-	objectACL.AccessControlList = append(objectACL.AccessControlList, AccessControlChange{
-		UserName:        me.UserName,
-		PermissionLevel: "CAN_MANAGE",
-	})
-	return objectACL, nil
+	return me.UserName, nil
 }
 
-// Helper function for applying permissions changes. Ensures that
-// we select the correct HTTP method based on the object type and preserve the calling
-// user's ability to manage the specified object when applying permissions changes.
-func (a PermissionsAPI) put(objectID string, objectACL AccessControlChangeList) error {
-	objectACL, err := a.ensureCurrentUserCanManageObject(objectID, objectACL)
+// Update updates object permissions. Technically, it's using method named SetOrDelete, but here we do more
+func (a PermissionsAPI) Update(objectID string, entity PermissionsEntity, mapping resourcePermissions) error {
+	currentUser, err := a.getCurrentUser()
 	if err != nil {
 		return err
 	}
-	if isDbsqlPermissionsWorkaroundNecessary(objectID) {
-		// SQLA entities use POST for permission updates.
-		return a.client.Post(a.context, urlPathForObjectID(objectID), objectACL, nil)
+	// this logic was moved from CustomizeDiff because of undeterministic auth behavior
+	// in the corner-case scenarios.
+	// see https://github.com/databricks/terraform-provider-databricks/issues/2052
+	err = mapping.validate(entity, currentUser)
+	if err != nil {
+		return err
 	}
-	log.Printf("[DEBUG] PUT %s %v", objectID, objectACL)
-	return a.client.Put(a.context, urlPathForObjectID(objectID), objectACL)
-}
-
-// safePutWithOwner is a workaround for the limitation where warehouse without owners cannot have IS_OWNER set
-func (a PermissionsAPI) safePutWithOwner(objectID string, objectACL AccessControlChangeList, originalAcl []AccessControlChange) error {
```

Comment:
Is this something we should also have in Go SDK and hence might be better to move it there?

Reply:
TBH I don't think it really belongs in the Go SDK. I would actually consider this more along the lines of an API issue for the permissions API for warehouses. This may not even be needed anymore now that warehouses and other SQL objects are using the global permissions API endpoints, but I need to test it out.